brian2/devices/cpp_standalone/device.py

```diff
@@ -568,18 +568,30 @@ def code_object(self, owner, name, abstract_code, variables, template_name,
         if self.enable_profiling:
             self.profiled_codeobjects.append(codeobj.name)
 
-        # Mark all the non-read-only or non-constant variables used in this code
-        # object as "dirty". This is almost certainly too much, most of these
-        # variables will only be read. However, the templates for synapse
-        # creation will write to read-only variables.. This is noted in the
-        # WRITES_TO_READ_ONLY_VARIABLES comment in the template.
+        # We mark all writeable (i.e. not read-only) variables used by the code
+        # as "dirty" to avoid that the cache contains incorrect values. This
+        # might remove a number of variables from the cache unnecessarily,
+        # since many variables are only read in the code.
+        # On the other hand, there are also *read-only* variables that can be
+        # changed by code (the "read-only" attribute only refers to the user
+        # being able to change values directly). For example, synapse creation
+        # write source and target indices, and monitors write the shared values.
+        # To correctly mark these values as changed, templates can include a
+        # "WRITES_TO_READ_ONLY_VARIABLES" comment, stating the name of the
+        # changed variables. For a monitor, this would for example state that
+        # the number of recorded values "N" changes. For the recorded variables,
+        # however, this information cannot be included in the template because
+        # it is up to the user to define which variables are recorded. For such
+        # cases, the "owner" object (e.g. a SpikeMonitor) can define a
+        # "written_readonly_vars" attribute, storing a set of `Variable` objects
+        # that will be changed by the owner's code objects.
```

Comment:
This is all getting deeply arcane and cements the idea that we need to do a refactoring at some point! :)

Reply:
I do agree! My long-term plan is that each code has explicit attributes stating which variables it reads and which it writes, which would not only make this stuff here obsolete but could also be used for optimisations such as running independent things in parallel.

Reply:
Yeah it feels like there's no reason why we shouldn't have all that information available in a clear and straightforward way given the general approach we're taking. Parallel idea is very neat!